src/Data/String/CodeUnits.ss

```diff
@@ -0,0 +1,128 @@
+;; -*- mode: scheme -*-
+
+(library (Data.String.CodeUnits foreign)
+  (export
+    fromCharArray
+    toCharArray
+    singleton
+    _charAt
+    _toChar
+    length
+    countPrefix
+    _indexOf
+    _indexOfStartingAt
+    _lastIndexOf
+    _lastIndexOfStartingAt
+    take
+    drop
+    slice
+    splitAt)
+  (import
+    (only (rnrs base) + < = > >= and cons define if lambda let max min or string string-length string-ref)
+    (prefix (purs runtime lib) rt:)
+    (prefix (purs runtime srfi :152) srfi:152:)
+    (prefix (purs runtime srfi :214) srfi:214:))
+
+  (define fromCharArray srfi:214:flexvector->string)
+
+  (define toCharArray srfi:214:string->flexvector)
+
+  (define singleton string)
+
+  (define _charAt
+    (lambda (just)
+      (lambda (nothing)
+        (lambda (i)
+          (lambda (s)
+            (if (and (>= i 0) (< i (string-length s)))
+              (just (string-ref s i))
+              nothing))))))
+
+  (define _toChar
+    (lambda (just)
+      (lambda (nothing)
+        (lambda (s)
+          (if (= 1 (string-length s))
+            (just (string-ref s 0))
+            nothing)))))
+
+  (define length string-length)
+
+  (define countPrefix
+    (lambda (p)
+      (lambda (s)
+        (or
+          (srfi:152:string-skip s p)
+          (string-length s)))))
+
+  (define _indexOf
+    (lambda (just)
+      (lambda (nothing)
+        (lambda (x)
+          (lambda (s)
+            (let ((i (srfi:152:string-contains s x)))
+              (if i (just i) nothing)))))))
+
+  (define _indexOfStartingAt
+    (lambda (just)
+      (lambda (nothing)
+        (lambda (x)
+          (lambda (startAt)
+            (lambda (s)
+              (if (or (> startAt (string-length s)) (< startAt 0))
+                nothing
+                (let ((i (srfi:152:string-contains s x startAt)))
+                  (if i (just i) nothing)))))))))
+
+  (define _lastIndexOf
+    (lambda (just)
+      (lambda (nothing)
+        (lambda (x)
+          (lambda (s)
+            (let ((i (srfi:152:string-contains-right s x)))
+              (if i (just i) nothing)))))))
+
+  (define _lastIndexOfStartingAt
+    (lambda (just)
+      (lambda (nothing)
+        (lambda (x)
+          (lambda (startAt)
+            (lambda (s)
+              (let ((clamped (clamp-to-length s (+ startAt (string-length x)))))
```

Comment:
I don't understand this clamping or why we need to do this. Shouldn't we just pass the `startAt` as-is to the `string-contains-right`?

Reply:
Is it because `startAt` can be negative?

Reply:
Clamping in general is needed because the library throws when index is outside the bounds. The weird `(+ startAt (string-length x))` thing is there because... it's easier to explain with an example from tests:
```
  assertEqual
    { actual: SCU.lastIndexOf' (Pattern "bc") 1 "abcd"
    , expected: Just 1
    }
```
I think it's more intuitive to expect it to return Nothing because if you start at index 1 and move left you won't ever get "bc", so you see the assumptions the JS/PS API makes about indexing here. That's why I add the length of the target pattern `x` to the starting index.

Reply:
Ok so `x` is not the whole string, it's the pattern. I missed that.